src/software/util/parameter/config_test.cpp

```diff
@@ -0,0 +1,241 @@
+                                  param->setValue(param->value() + "test");
+                              },
+                              [&](std::shared_ptr<Parameter<double>> param) {
+                                  param->setValue(param->value() - 2.0);
+                              },
+                              [&](std::shared_ptr<Config> param) {
+                                  for (auto& v : param->getMutableParameterList())
+                                  {
+                                      mutate_all_parameters(v);
+                                  }
+                              }},
+                   paramvar);
+    }
+
+    /*
+     * Tests that the mutations in the mutable parameter list tree are propagated to
+     * the immutable tree
+     *
+     * NOTE: at any given time this test may fail due to the key not existing in the yaml
+     * dictionary which indicates an error w/ the generate script.
+     */
+    void assert_mutation(ParameterVariant paramvar, const YAML::Node& current_config)
+    {
+        std::visit(
+            overloaded{
+                [&](std::shared_ptr<const Parameter<int>> param) {
+                    ASSERT_EQ(current_config[param->name()]["default"].as<int>(),
+                              param->value() - 4);
+                },
+                [&](std::shared_ptr<const Parameter<bool>> param) {
+                    ASSERT_EQ(current_config[param->name()]["default"].as<bool>(),
+                              !param->value());
+                },
+                [&](std::shared_ptr<const Parameter<std::string>> param) {
+                    ASSERT_EQ(current_config[param->name()]["default"].as<std::string>() +
+                                  "test",
+                              param->value());
+                },
+                [&](std::shared_ptr<const Parameter<double>> param) {
+                    ASSERT_NEAR(current_config[param->name()]["default"].as<double>(),
+                                param->value() + 2.0, 1E-10);
+                },
+                [&](std::shared_ptr<const Config> param) {
+                    for (auto& v : param->getParameterList())
+                    {
+                        assert_mutation(v, current_config[param->name()]);
+                    }
+                }},
+            paramvar);
+    }
+};
+
+TEST_F(TestParameterMutation, DynamicParametersTest)
+{
+    // This creates a shared ptr pointing to a ThunderbotsConfig which can be mutated
+    const std::shared_ptr<ThunderbotsConfig> MutableDynamicParameters =
+        std::make_shared<ThunderbotsConfig>();
+
+    // This creates an immutable ThunderbotsConfig w/ proper const correctnesss
+    const std::shared_ptr<const ThunderbotsConfig> DynamicParameters =
+        std::const_pointer_cast<const ThunderbotsConfig>(MutableDynamicParameters);
+
+    mutate_all_parameters(MutableDynamicParameters);
+    assert_mutation(DynamicParameters, config_yaml);
+}
```

Comment:
This is super slick :heavy_check_mark:

Reply:
lol just wanted to 🅱️e safe

Reply:
💯

Reply:
Again with that :b:. I swear I'm not :older_man: but I still don't get it :confused: